internal/crypto/internal_ca.go

```diff
@@ -0,0 +1,160 @@
+		SerialGenerator: serialGenerator,
+		Config:          &config,
+	}, nil
+}
+
+func makeSelfSignedCAConfig(subject pkix.Name, caLifetime time.Duration) (*oscrypto.TLSCertificateConfig, error) {
+	rootcaPublicKey, rootcaPrivateKey, publicKeyHash, err := NewKeyPairWithHash()
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now()
+	rootcaTemplate := &x509.Certificate{
+		Subject: subject,
+
+		SignatureAlgorithm: x509.ECDSAWithSHA256,
+
+		NotBefore: now.Add(-1 * time.Second),
+		NotAfter:  now.Add(caLifetime),
+
+		SerialNumber: randomSerial(),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+
+		AuthorityKeyId: publicKeyHash,
+		SubjectKeyId:   publicKeyHash,
+	}
+	rootcaCert, err := signCertificate(rootcaTemplate, rootcaPublicKey, rootcaTemplate, rootcaPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+	caConfig := &oscrypto.TLSCertificateConfig{
+		Certs: []*x509.Certificate{rootcaCert},
+		Key:   rootcaPrivateKey,
+	}
+	return caConfig, nil
+}
+
+func (caBackend *internalCA) signCertificate(template *x509.Certificate, requestKey crypto.PublicKey) (*x509.Certificate, error) {
+	// Increment and persist serial
+	serial, err := caBackend.SerialGenerator.Next(template)
+	if err != nil {
+		return nil, err
+	}
+	template.SerialNumber = big.NewInt(serial)
+	return signCertificate(template, requestKey, caBackend.Config.Certs[0], caBackend.Config.Key)
+}
+
+func signCertificate(template *x509.Certificate, requestKey crypto.PublicKey, issuer *x509.Certificate, issuerKey crypto.PrivateKey) (*x509.Certificate, error) {
+	derBytes, err := x509.CreateCertificate(rand.Reader, template, issuer, requestKey, issuerKey)
+	if err != nil {
+		return nil, err
+	}
+	certs, err := x509.ParseCertificates(derBytes)
+	if err != nil {
+		return nil, err
+	}
+	if len(certs) != 1 {
+		return nil, errors.New("expected a single certificate")
+	}
+	return certs[0], nil
+}
+
+// IssueRequestedClientCertificate issues a client certificate based on the provided
+// Certificate Signing Request (CSR) and the desired expiration time in seconds.
+// This currently processes both enrollment cert and management cert signing requests, which both are signed
+// by the FC service's internal CA instance named 'ca'.
+func (caBackend *internalCA) IssueRequestedCertificateAsX509(csr *x509.CertificateRequest, expirySeconds int, usage []x509.ExtKeyUsage) (*x509.Certificate, error) {
+	now := time.Now()
+	expire := time.Duration(expirySeconds) * time.Second
+	template := &x509.Certificate{
+		Subject:               csr.Subject,
+		Signature:             csr.Signature,
+		SignatureAlgorithm:    csr.SignatureAlgorithm,
+		PublicKey:             csr.PublicKey,
+		PublicKeyAlgorithm:    csr.PublicKeyAlgorithm,
+		IPAddresses:           csr.IPAddresses,
+		DNSNames:              csr.DNSNames,
+		Issuer:                caBackend.Config.Certs[0].Subject,
```

Comment:
can we add a note that the length is ensured by signCertificate or would it make sense to ensure here

Reply:
Ack